File: spiralMat.py
# // Time Complexity : O(m*n) where m*n: number of elements in the matrix
# // Space Complexity : O(m*n) where m*n: number of elements
# // Did this code successfully run on Leetcode : yes
# // Any problem you faced while coding this : none

# // Your code here along with comments explaining your approach
class Solution:
    def spiralOrder(self, matrix: List[List[int]]) -> List[int]:
        row_start = 0
        row_end = len(matrix) - 1
        col_start = 0
        col_end = len(matrix[0]) - 1
        res = []
        
        while row_start <= row_end and col_start <= col_end:
            
            # top row
            for c in range(col_start,col_end+1):
                res.append(matrix[row_start][c])
            row_start += 1
            
            if row_start > row_end: break
                
            # last col
            for r in range(row_start,row_end+1):
                res.append(matrix[r][col_end])
            col_end -= 1
            
            if col_end < col_start: break
                
            # last row
            if row_start <= row_end:
                
                for c in range(col_end,col_start-1,-1):
                    res.append(matrix[row_end][c])
                row_end -= 1
            
            # first col
            if col_start <= col_end:
                
                for r in range(row_end,row_start-1,-1):
                    res.append(matrix[r][col_start])
                col_start += 1
            
        return res
        

File: SpiralMat.java
class Solution {
  #time complexity = O(m*n)
  #space complexity = O(1)
  
  public List<Integer> spiralOrder(int[][] mat) {
        int n = mat.length;
        int m = mat[0].length;
        List<Integer> res = new ArrayList<>();
        int left =0; int right = m-1;
        int top =0; int bot = n-1;
        while(left<=right && top<=bot){
            
            for(int i = left; i<=right; i++){
                res.add(mat[top][i]);
            }
            top++;
            if(left<=right){
                for(int i = top; i<=bot; i++){
                    res.add(mat[i][right]);
                }
                right--;
            }
            if(top<=bot){
                for(int i = right; i>=left; i--){
                    res.add(mat[bot][i]);
                }
                bot--;
            }
            if(left<=right){
                for(int i = bot; i>=top; i--){
                res.add(mat[i][left]);
                }
                left++;
            }
            
        }
        return res;
    }
}
